File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/world-events.ts
/**
 * Deterministic campaign-level world events, scheduled by the DM for a
 * future campaign turn and resolved exactly once when that turn arrives.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

type WorldEventRow = {
  event_id: string;
  turn_number: number;
  title: string;
  text: string;
  status: string;
  resolution_turn_number: number | null;
  resolution_text: string | null;
};

const WORLD_EVENT_COLUMNS =
  'event_id, turn_number, title, text, status, resolution_turn_number, resolution_text';

function worldEventBody(row: WorldEventRow): JsonValue {
  const body: Record<string, unknown> = {
    event_id: row.event_id,
    turn_number: row.turn_number,
    title: row.title,
    text: row.text,
    status: row.status,
  };
  if (row.status === 'resolved') {
    body.resolution = { turn_number: row.resolution_turn_number, text: row.resolution_text };
  }
  return body as JsonValue;
}

function nextWorldEventSequence(db: ReturnType<typeof getDb>, campaignId: string): number {
  const row = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_world_events WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };
  return row.max_sequence + 1;
}

export function scheduleWorldEvent(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may schedule world events' } };
  }

  const eventId = body.event_id;
  if (typeof eventId !== 'string' || eventId.length === 0) {
    return { status: 400, body: { error: 'event_id must be a non-empty string' } };
  }
  const title = body.title;
  if (typeof title !== 'string' || title.length === 0) {
    return { status: 400, body: { error: 'title must be a non-empty string' } };
  }
  const text = body.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }
  const turnNumber = body.turn_number;
  const currentTurn = campaign.turn_number ?? 0;
  if (
    typeof turnNumber !== 'number' ||
    !Number.isInteger(turnNumber) ||
    turnNumber < currentTurn
  ) {
    return { status: 400, body: { error: 'turn_number must be an integer at or after the current turn' } };
  }

  const existing = db
    .prepare('SELECT event_id FROM play_campaign_world_events WHERE campaign_id = ? AND event_id = ?')
    .get(campaignId, eventId);
  if (existing) {
    return { status: 409, body: { error: 'event_id already exists in this campaign' } };
  }

  const sequence = nextWorldEventSequence(db, campaignId);
  db.prepare(
    'INSERT INTO play_campaign_world_events (campaign_id, sequence, event_id, turn_number, title, text, status) VALUES (?, ?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequence, eventId, turnNumber, title, text, 'scheduled');

  return {
    status: 201,
    body: worldEventBody({
      event_id: eventId,
      turn_number: turnNumber,
      title,
      text,
      status: 'scheduled',
      resolution_turn_number: null,
      resolution_text: null,
    }),
  };
}

export function resolveWorldEvent(
  authHeader: string | undefined,
  campaignId: string,
  eventId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may resolve world events' } };
  }

  const eventRow = db
    .prepare(`SELECT ${WORLD_EVENT_COLUMNS} FROM play_campaign_world_events WHERE campaign_id = ? AND event_id = ?`)
    .get(campaignId, eventId) as WorldEventRow | undefined;
  if (!eventRow) {
    return { status: 404, body: { error: 'world event not found' } };
  }

  const resolutionText = body.text;
  if (typeof resolutionText !== 'string' || resolutionText.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }

  if (eventRow.status === 'resolved') {
    return { status: 409, body: { error: 'world event has already been resolved' } };
  }

  const currentTurn = campaign.turn_number ?? 0;
  if (currentTurn !== eventRow.turn_number) {
    return { status: 409, body: { error: 'campaign turn number does not match the event turn number' } };
  }

  db.prepare(
    'UPDATE play_campaign_world_events SET status = ?, resolution_turn_number = ?, resolution_text = ? WHERE campaign_id = ? AND event_id = ?',
  ).run('resolved', eventRow.turn_number, resolutionText, campaignId, eventId);

  return {
    status: 201,
    body: worldEventBody({
      ...eventRow,
      status: 'resolved',
      resolution_turn_number: eventRow.turn_number,
      resolution_text: resolutionText,
    }),
  };
}

export function listWorldEvents(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const rows = db
    .prepare(
      `SELECT ${WORLD_EVENT_COLUMNS} FROM play_campaign_world_events WHERE campaign_id = ? ORDER BY turn_number ASC, sequence ASC`,
    )
    .all(campaignId) as WorldEventRow[];

  return { status: 200, body: { events: rows.map((row) => worldEventBody(row)) } };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/world_events.rb
# Deterministic campaign-level world events scheduled by the DM for a
# future campaign turn and resolved exactly once that turn is reached.

def play_world_event_payload(event)
  payload = {
    event_id: event['event_id'],
    turn_number: event['turn_number'],
    title: event['title'],
    text: event['text'],
    status: event['status']
  }

  if event['status'] == 'resolved'
    payload[:resolution] = {
      turn_number: event['resolution_turn_number'],
      text: event['resolution_text']
    }
  end

  payload
end

post '/v1/play/campaigns/:id/world-events' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  event_id = body['event_id']
  turn_number = body['turn_number']
  title = body['title']
  text = body['text']

  halt 400, { error: 'invalid event_id' }.to_json unless event_id.is_a?(String) && !event_id.empty?
  halt 400, { error: 'invalid title' }.to_json unless title.is_a?(String) && !title.empty?
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?
  halt 400, { error: 'invalid turn_number' }.to_json unless integerish(turn_number)
  turn_number = turn_number.to_i
  halt 400, { error: 'invalid turn_number' }.to_json if turn_number < campaign['turn_number'].to_i

  existing = db.execute(
    'SELECT 1 FROM play_world_events WHERE campaign_id = ? AND event_id = ?',
    [campaign['id'], event_id]
  ).first
  halt 409, { error: 'world event already exists' }.to_json if existing

  sequence = db.execute(
    'SELECT COALESCE(MAX(sequence), -1) + 1 AS seq FROM play_world_events WHERE campaign_id = ?',
    [campaign['id']]
  ).first['seq']

  db.execute(
    'INSERT INTO play_world_events (campaign_id, sequence, event_id, turn_number, title, text, status) VALUES (?, ?, ?, ?, ?, ?, ?)',
    [campaign['id'], sequence, event_id, turn_number, title, text, 'scheduled']
  )

  status 201
  { event_id: event_id, turn_number: turn_number, title: title, text: text, status: 'scheduled' }.to_json
end

post '/v1/play/campaigns/:id/world-events/:event_id/resolve' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  event = db.execute(
    'SELECT * FROM play_world_events WHERE campaign_id = ? AND event_id = ?',
    [campaign['id'], params[:event_id]]
  ).first
  halt 404, { error: 'world event not found' }.to_json unless event

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  text = body['text']
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?

  halt 409, { error: 'world event already resolved' }.to_json if event['status'] == 'resolved'
  halt 409, { error: 'turn mismatch' }.to_json unless campaign['turn_number'].to_i == event['turn_number']

  db.execute(
    'UPDATE play_world_events SET status = ?, resolution_turn_number = ?, resolution_text = ? WHERE campaign_id = ? AND event_id = ?',
    ['resolved', event['turn_number'], text, campaign['id'], params[:event_id]]
  )

  event['status'] = 'resolved'
  event['resolution_turn_number'] = event['turn_number']
  event['resolution_text'] = text

  status 201
  play_world_event_payload(event).to_json
end

get '/v1/play/campaigns/:id/world-events' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  events = db.execute(
    'SELECT * FROM play_world_events WHERE campaign_id = ? ORDER BY turn_number ASC, sequence ASC',
    [campaign['id']]
  )

  { events: events.map { |event| play_world_event_payload(event) } }.to_json
end
